autoRetryFailedCRR-S3C.js

```diff
@@ -0,0 +1,276 @@
+            if (!objMD.replicationInfo ||
+                objMD.replicationInfo.status !== 'FAILED') {
+                log.info('skipping object: not FAILED', {
+                    bucket,
+                    key,
+                    versionId,
+                    status: objMD.replicationInfo ?
+                        objMD.replicationInfo.status : 'NEW',
+                });
+                ++counters.skipped;
+                return next();
+            }
+            const objSize = objMD['content-length'];
+            // reset to PENDING
+            objMD.replicationInfo.status = 'PENDING';
+            objMD.replicationInfo.backends[0].status = 'PENDING';
+            const mdBlob = JSON.stringify(objMD);
+            // create an entry as if coming from the raft log
+            const entry = JSON.stringify({
+                type: 'put',
+                bucket,
+                key: `${key}${VID_SEP}${objMD.versionId}`,
+                value: mdBlob,
+            });
+            return retry({
+                times: PRODUCER_MAX_RETRIES,
+                interval: PRODUCER_RETRY_DELAY_MS,
+            }, attemptDone => {
+                try {
+                    producer.produce(
+                        KAFKA_TOPIC,
+                        null, // partition
+                        new Buffer(entry), // value
+                        `${bucket}/${key}`, // key (for keyed partitioning)
+                        Date.now(), // timestamp
+                        null);
+                    log.info('requeued object version for replication', {
+                        bucket,
+                        key,
+                        versionId,
+                    });
+                    ++counters.requeued;
+                    counters.requeuedBytes += objSize;
+                    return attemptDone();
+                } catch (err) {
+                    log.error('error producing entry to kafka, retrying', {
+                        bucket,
+                        key,
+                        error: err.message,
+                    });
+                    return attemptDone(err);
+                }
+            }, err => {
+                if (err) {
+                    log.error('give up producing entry to kafka after retries',
+                              { bucket, key, error: err.message });
+                    ++counters.errors;
+                }
+                next(err);
+            });
+        },
+    ], cb);
+    /* eslint-enable no-param-reassign */
+}
+
+function _requeueAll() {
+    log.info('starting requeuing task');
+    const counters = {
+        requeued: 0,
+        requeuedBytes: 0,
+        skipped: 0,
+        errors: 0,
+    };
+    let marker = undefined;
+    requeueInProgress = true;
+    doWhilst(
+        batchDone => {
+            log.info('requeuing task progress', counters);
+            let url = `${BACKBEAT_API_ENDPOINT}/_/crr/failed?sitename=${SITE_NAME}`;
```

Comment:
Not a review comment, but a question, what does this API return?

Reply:
If it does, then do we need this check https://github.com/scality/s3utils/pull/117/files#diff-cb218756cc48d39100387adb8eab0967e440a4994b72fc86f7c8800eb8cf0860R103?

Reply:
Yes, something like:

```
{ "Versions": [{ "Bucket": "foo", "Key": "bar", "VersionId": "42" }, {...}], "IsTruncated": false }
```

Reply:
I'm considering that the object can have any replication status. At least it might not be `FAILED` because we do not clear this list when requeueing the object. I don't think it can be "new" i.e. with no replication status (or replication info) at all if it is in the list, but in doubt I think leaving this check does not hurt.

Reply:
my second question :up:

Reply:
looks good to me.

Reply:
answered :up: :laughing: